### PR TITLE
[Feature] DCP-aware CPU KV offload for the vLLM v1 connector

### PR DESCRIPTION
**Pull request by RomaA2000**
**What this PR does / why we need it**

Enables LMCache CPU KV offload under vLLM decode-context-parallel (DCP, `decode_context_parallel_size > 1`). The v1 connector assumes the MLA latent KV is replicated, so under DCP, where each rank holds only a strided `1/cp_world` slice of the context, the global save rank never holds a complete prefix. The store either asserts (`len(slot_mapping) == len(token_ids)`) or no-ops, and LMCache gets zero CPU hits under DCP.

This adds the missing DCP coupling in a self-contained module (`lmcache/integration/vllm/dcp_gather.py`) wired into the connector with a small guarded branch in `wait_for_save` and `start_load_kv`. The non-DCP path is untouched because the branch returns `False` and the normal code runs.

On SAVE each rank reads its shard straight from the paged KV caches, all-gathers across the DCP group, block-interleaves the shards into the full chunk, and the save rank stores it under the full-token key so the scheduler's full-token lookup matches and hits.

On LOAD the save rank fetches the full chunk, broadcasts it across the TP group, and each rank block-deinterleaves its own shard back into its paged caches.

The paged caches are indexed and scattered directly with no per-rank `MemoryObj`, so the non-save ranks, which have no CPU cache under `save_only_first_rank`, still participate in the collectives while only the save rank touches the cache.

**Validation**

Validated on B200 (SM100), TP=4, Kimi-K2.6-NVFP4, fp8 KV cache, eagle3 spec, on a Mooncake agentic-trace replay. Correctness is gated by IFEval through the aligned lm_eval harness with reasoning disabled, and the offload is output-neutral, matching the no-offload baseline at DCP=2 (prompt_strict 0.88) and DCP=4 (prompt_strict 0.89).

CPU offload is confirmed active under DCP with 30k+ hit-events and single-request CPU loads into the hundreds of thousands of tokens, with no crashes and output lengths identical to the no-offload run.

One serving-side requirement worth calling out for reviewers reproducing this: the DCP run must use the strided KV layout (`cp_kv_cache_interleave_size=1`).

**If applicable**

- [x] this PR contains unit tests (`tests/v1/test_dcp_gather.py`, the interleave and deinterleave round-trip plus strided-block ownership, CPU-only with no GPU or distributed dependency)
- [ ] this PR contains user facing changes, the behavior is automatic under DCP and the design and layout requirement are documented in the module docstring

### CODE REVIEW

File: lmcache/integration/vllm/dcp_gather.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+collective + copy traffic. The bottleneck is the data movement itself, not the
+number of collectives: coalescing the per-chunk collectives into batched ones and
+skipping the replica DCP group's gather were both measured and *regressed*
+throughput (extra cat/slice copies and larger synchronous broadcasts inflate
+TTFT). The real win is to overlap the load with prefill compute (async prefetch /
+double-buffering) so it leaves the critical path -- left as future work.
+"""
+
+# Standard
+from typing import TYPE_CHECKING
```
[the remark this answers is not in the document]
Comment (RomaA2000) [reply lmcache/integration/vllm/dcp_gather.py]: Updated

File: lmcache/integration/vllm/dcp_gather.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+
+# Third Party
+import torch
+
+# First Party
+from lmcache.logging import init_logger
+
+if TYPE_CHECKING:
+    # First Party
+    from lmcache.integration.vllm.vllm_v1_adapter import LMCacheConnectorV1Impl
```
[the remark this answers is not in the document]
Comment (RomaA2000) [reply lmcache/integration/vllm/dcp_gather.py]: Updated

File: lmcache/integration/vllm/dcp_gather.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+
+def _flat(layer_cache: torch.Tensor) -> torch.Tensor:
+    # Paged latent cache [num_blocks, block_size, (1,) D] -> VIEW
+    # [num_blocks * block_size, D]. .view (not .reshape) so index_copy_ on the load
+    # path writes THROUGH to the real cache; raises if it is ever non-contiguous
+    # (fail fast rather than silently scatter into a copy).
+    return layer_cache.view(-1, layer_cache.shape[-1])
+
+
+def _extract_shard(kvcaches, rslots: torch.Tensor) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (RomaA2000) [reply lmcache/integration/vllm/dcp_gather.py]: Updated

File: lmcache/integration/vllm/dcp_gather.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+    try:
+        _, world, _ = _dcp_group()
+    except Exception:
+        return False
+    return world > 1 and not impl.use_layerwise and impl.kv_role != "kv_consumer"
+
+
+def maybe_dcp_save(impl: "LMCacheConnectorV1Impl", connector_metadata) -> bool:
+    """If DCP is active, gather+store every request's KV and return True; else
+    return False so the caller runs the normal (non-DCP) save."""
```
[the remark this answers is not in the document]
Comment (RomaA2000) [reply lmcache/integration/vllm/dcp_gather.py]: Updated

File: lmcache/integration/vllm/dcp_gather.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+    if not dcp_gather_enabled(impl):
+        return False
+    kvcaches = list(impl.kv_caches.values())
+    assert len(kvcaches) > 0
+    dev = kvcaches[0].device
+    chunk = impl._lmcache_chunk_size
```
[the remark this answers is not in the document]
Comment (RomaA2000) [reply lmcache/integration/vllm/dcp_gather.py]: Updated

File: lmcache/integration/vllm/dcp_gather.py [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+        if save_spec is not None:
+            save_spec.skip_leading_tokens = len(token_ids)
+    if total:
+        logger.info("DCP gather: stored %d chunks across DCP ranks", total)
+    return True
+
+
+def maybe_dcp_load(impl: "LMCacheConnectorV1Impl", connector_metadata) -> bool:
+    """If DCP is active, load+scatter every request's KV and return True; else
+    return False so the caller runs the normal (non-DCP) load."""
```
[the remark this answers is not in the document]
Comment (RomaA2000) [reply lmcache/integration/vllm/dcp_gather.py]: Updated

File: lmcache/integration/vllm/dcp_gather.py [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+    if not dcp_gather_enabled(impl):
+        return False
+    kvcaches = list(impl.kv_caches.values())
+    dev = kvcaches[0].device
+    chunk = impl._lmcache_chunk_size
```
[the remark this answers is not in the document]
Comment (RomaA2000) [reply lmcache/integration/vllm/dcp_gather.py]: Updated

File: lmcache/integration/vllm/dcp_gather.py [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+    total = 0
+    for request in connector_metadata.requests:
+        load_spec = request.load_spec
+        if load_spec is not None:
+            impl._stats_monitor.update_interval_vllm_hit_tokens(
+                load_spec.vllm_cached_tokens
+            )
+            impl._stats_monitor.update_interval_prompt_tokens(len(request.token_ids))
```
[the remark this answers is not in the document]
Comment (RomaA2000) [reply lmcache/integration/vllm/dcp_gather.py]: Updated

File: lmcache/integration/vllm/dcp_gather.py [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+    return layer_cache.view(-1, layer_cache.shape[-1])
+
+
+def _extract_shard(kvcaches, rslots: torch.Tensor) -> torch.Tensor:
+    # This rank's latent KV for `rslots` -> [1, L, nl, D] (KV_MLA_FMT, kv_size=1).
+    parts = [_flat(c).index_select(0, rslots) for c in kvcaches]  # each [nl, D]
+    return torch.stack(parts, dim=0).unsqueeze(0)  # [1, L, nl, D]
+
+
+def _scatter_shard(kvcaches, rslots: torch.Tensor, shard: torch.Tensor) -> None:
```
[the remark this answers is not in the document]
Comment (RomaA2000) [reply lmcache/integration/vllm/dcp_gather.py]: Updated

File: lmcache/integration/vllm/dcp_gather.py [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+            full.tensor.copy_(_interleave(gathered, world, blk, n))  # GPU -> CPU obj
+            keys.append(key)
+            mobjs.append(full)
+        n_chunks += 1
+    if storage is not None and keys:
+        storage.batched_put(keys, mobjs, location=engine.store_location)
+    return n_chunks
+
+
+def _dcp_load(impl, token_ids, token_mask, kvcaches, slot_mapping, n_load) -> int:
```
[the remark this answers is not in the document]
Comment (RomaA2000) [reply lmcache/integration/vllm/dcp_gather.py]: Updated

File: lmcache/integration/vllm/dcp_gather.py [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+            mobj.ref_count_down()
+        tp_group.broadcast(full_gpu, src=0)  # full prefix KV to every TP rank
+        rslots = slot_mapping[start // world : end // world]
+        _scatter_shard(kvcaches, rslots, _deinterleave(full_gpu, world, rank, blk, nl))
+        loaded += n
+    return loaded
+
+
+def dcp_gather_enabled(impl: "LMCacheConnectorV1Impl") -> bool:
+    """Whether this connector should use the DCP gather/scatter offload path."""
```
[the remark this answers is not in the document]
Comment (RomaA2000) [reply lmcache/integration/vllm/dcp_gather.py]: Updated

File: lmcache/integration/vllm/dcp_gather.py [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+    kv, length, _, dim = full_t.shape
+    nlb = nl // blk
+    return (
+        full_t.view(kv, length, nlb, world, blk, dim)[:, :, :, rank, :, :]
+        .reshape(kv, length, nl, dim)
+        .contiguous()
+    )
+
+
+def _dcp_store(impl, token_ids, store_mask, kvcaches, slot_mapping) -> int:
```
[the remark this answers is not in the document]
Comment (RomaA2000) [reply lmcache/integration/vllm/dcp_gather.py]: Updated

File: lmcache/integration/vllm/dcp_gather.py [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+    # First Party
+    from lmcache.integration.vllm.vllm_v1_adapter import LMCacheConnectorV1Impl
+
+logger = init_logger(__name__)
+
+_blk_size: "int | None" = None
+_shape_logged = False
+
+
+def _dcp_group():
```
[the remark this answers is not in the document]
Comment (RomaA2000) [reply lmcache/integration/vllm/dcp_gather.py]: Updated

File: lmcache/integration/vllm/dcp_gather.py [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+"""
+
+# Standard
+from typing import TYPE_CHECKING
+
+# Third Party
+import torch
+
+# First Party
+from lmcache.logging import init_logger
```
[the remark this answers is not in the document]
Comment (RomaA2000) [reply lmcache/integration/vllm/dcp_gather.py:46]: Updated

File: lmcache/integration/vllm/dcp_gather.py [hunk 15]
```diff
@@ -0,0 +1,351 @@
...
+
+def _dcp_group() -> "tuple[Any, int, int]":
+    # Third Party
+    from vllm.distributed.parallel_state import get_dcp_group
+
+    group = get_dcp_group()
+    return group, group.world_size, group.rank_in_group
+
+
+def _interleave_size() -> int:
```
Comment (sammshen) [inline lmcache/integration/vllm/dcp_gather.py:70]: your comment says this requires `cp_kv_cache_interleave_size=1` the fallback to blk=64 produces wrong KV with no crash

Comment (RomaA2000) [reply lmcache/integration/vllm/dcp_gather.py:70]: Fixed. `_interleave_size()` now raises a clear `ValueError` when `cp_kv_cache_interleave_size != 1` (and defaults to 1, not 64, when the config is unreadable), so a block-granular layout fails fast instead of silently storing/scattering wrong KV — the strided requirement is enforced in code now, not just documented. Also merged latest `dev`.
